formal-spec/update-protocol.tex

```diff
@@ -0,0 +1,99 @@
+  used to supersede a proposal that is already in the activation phase. This
+  allows experts pools to overrule a previously approved proposal if a problem
+  is detected after it enters the activation phase.
+\end{itemize}
+
+When the implementation of a proposal gets approved a \emph{conflict resolution}
+takes place:
+\begin{itemize}
+\item If there are approved proposals (either improvement proposals or
+  implementations of them) that conflict with the newly approved proposal, we
+  remove them from the set of approved proposals.
+\item If there is a proposal in the activation phase that conflicts with the
+  newly approved proposal, the former is removed from the activation phase. This
+  mechanism allows expert pools to quickly react to potential errors in an
+  implementation that entered the activation phase, which were discovered after
+  the approval of said implementation.
+\end{itemize}
+
+Approved implementations enter the activation phase based on their
+\emph{priority} and order in a \emph{priority queue}. Priority is specified via
+the the urgency of a proposal:
+\begin{enumerate}
+\item critical: delaying this could cause a major disruption in the chain.
+\item hotfix: fix that should be activated as soon as possible, but whose
+  activation can wait in case there are critical updates.
+\item normal: planned chain updates.
+\end{enumerate}
+In principle, it seems there is no technical reason for having proposals being
+sorted by their versions, but we can add this as second ordering criteria.
+Adding this criteria might be desirable since users will see protocol versions
+increase monotonically, rather than jumping arbitrarily between updates.
+
+Whenever a new proposal can enter the activation phase we traverse the
+priority queue of proposals till one is found for which \emph{all of its
+  dependencies are met}.
+
+\subsubsection{On cut-off points for endorsements}
+\label{sec:on-cutoff-points-for-endorsements}
+
+The ledger layer must provide future information about a part of its state. In
+particular, it should be able to tell up to $2k$ slots in the future if an new
+protocol version will be activated. To this end, we require that the \emph{last
+  endorsement} required for meeting the activation threshold \emph{is stable}
+$2k$ slots before the end of the epoch. This means that endorsements for a given
+proposal for a given epoch are considered up to $4k$ slots before the end of
+that epoch. This is not to say that the endorsements after this cut-off point
+will not be considered, just that they will in the next epoch.
+
+To see why this is required, consider a last endorsement (required to meet the
+threshold) arriving between $4k$ slots and $2k$ slots before the end of an
+epoch, which will happen at slot $s_e$. If the ledger is asked about whether a
+new version will be activated between $s_e - 2k$ and $s_e$, the answer depends
+on the stability of the last endorsement. If it is stable, then replying ``yes''
```

Comment:
"If it is stable"  -> Maybe you wanted to say "if it is not stable"

Reply:
No I think this is correct. However I think what I wrote is wrong: if the endorsement is stable, when we're fine: even if we roll back and switch to a different fork, we won't be able to switch to a fork in which the endorsement didn't occur. The only problem is if the endorsement is not stable.